app/services/data_decoder.py

```diff
@@ -0,0 +1,393 @@
+import logging
+from functools import cache
+from typing import Any, AsyncIterator, NotRequired, TypedDict, Union, cast
+
+from async_lru import alru_cache
+from eth_abi import decode as decode_abi
+from eth_abi.exceptions import DecodingError
+from eth_typing import Address, ChecksumAddress, HexStr
+from eth_utils import function_abi_to_4byte_selector
+from hexbytes import HexBytes
+from safe_eth.eth.contracts import get_multi_send_contract
+from safe_eth.safe.multi_send import MultiSend
+from sqlmodel.ext.asyncio.session import AsyncSession
+from web3 import Web3
+from web3._utils.abi import get_abi_input_names, get_abi_input_types, map_abi_data
+from web3._utils.normalizers import BASE_RETURN_NORMALIZERS
+from web3.types import ABI, ABIFunction
+
+from ..datasources.db.database import get_engine
+from ..datasources.db.models import Abi, Contract
+
+logger = logging.getLogger(__name__)
+
+
+class DataDecoderException(Exception):
+    pass
+
+
+class UnexpectedProblemDecoding(DataDecoderException):
+    pass
+
+
+class CannotDecode(DataDecoderException):
+    pass
+
+
+class ParameterDecoded(TypedDict):
+    name: str
+    type: str
+    value: Any
+    value_decoded: NotRequired[Union[list["MultisendDecoded"], "DataDecoded", None]]
+
+
+class DataDecoded(TypedDict):
+    method: str
+    parameters: list[ParameterDecoded]
+
+
+class MultisendDecoded(TypedDict):
+    operation: int
+    to: ChecksumAddress
+    value: str
+    data: HexStr | None
+    data_decoded: DataDecoded | None
+
+
+@cache
+async def get_data_decoder_service() -> "DataDecoderService":
+    async with AsyncSession(get_engine(), expire_on_commit=False) as session:
+        data_decoder_service = DataDecoderService()
+        await data_decoder_service.init(session)
+    return data_decoder_service
+
+
+class DataDecoderService:
+    EXEC_TRANSACTION_SELECTOR = HexBytes("0x6a761202")
+
+    dummy_w3 = Web3()
+    session: AsyncSession | None
+
+    async def init(self, session: AsyncSession) -> None:
+        """
+        Initialize the data decoder service, loading the ABIs from the database and storing the 4byte selectors
+        in memory
+        """
+        logger.info("%s: Loading contract ABIs for decoding", self.__class__.__name__)
+        self.fn_selectors_with_abis: dict[bytes, ABIFunction] = (
+            await self._generate_selectors_with_abis_from_abis(
```

Comment:
Would make sense add this as a field for the ABI? Not in this PR but for the future.